source/common/network/listener_impl.cc

```diff
@@ -37,11 +41,11 @@ void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr*
     // Hands off redirected connections (from iptables) to the listener associated with the
     // original destination address. If there is no listener associated with the original
     // destination address, the connection is handled by the listener that receives it.
-    // Note: A listener that has the use_original_dst flag set to true can still receive.
+    // Note: A listener that has the use_original_dst flag set to true can still receive
     // connections that are NOT redirected using iptables. If a connection was not redirected,
-    // the address returned by getOriginalDst() match the listener address. In this case the
-    // listener handles the connection directly and does not hand it off.
-    if (listener->socket_.localAddress() != final_local_address) {
+    // the address returned by getOriginalDst() matches the local address of the new socket.
+    // In this case the listener handles the connection directly and does not hand it off.
+    if (!(*final_local_address == *local_address)) {
```

Comment:
I think the latter is fine in this case. FYI I need this in a PR that I'm working on now, so will add it to my PR and whichever merges first can add it in.